// libsvm-ts/build/SVM.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-ignore: Emscripten starting point ASM
var libsvm_1 = __importDefault(require("../out/asm/libsvm"));
// @ts-ignore: Emscripten starting point for WASM
var libsvm_2 = __importDefault(require("../out/wasm/libsvm"));
var Errors_1 = require("./Errors");
var Commands_1 = require("./types/Commands");
var Util_1 = require("./Util");
var SVM = /** @class */ (function () {
    function SVM(options) {
        if (Commands_1.SVMTypes.indexOf(options.type) === -1) {
            throw Errors_1.SVMError("SVM cannot instantiate with an unknown type " + options.type);
        }
        if (Commands_1.KernelTypes.indexOf(options.kernel) === -1) {
            throw Errors_1.SVMError("SVM cannot instantiate with an unknown kernel " + options.kernel);
        }
        this.options = options;
        this.model = null;
        this.loaded = false;
    }
    SVM.prototype.loadASM = function () {
        var _this = this;
        return libsvm_1.default
            .load()
            .then(function () {
            if (_this.loaded) {
                throw Errors_1.SVMError('Cannot load an already loaded SVM');
            }
            _this.libsvm = libsvm_1.default;
            _this.initiateAPIs(_this.libsvm);
            _this.loaded = true;
            return _this;
        })
            .catch(function (err) {
            throw Errors_1.WASMError(err);
        });
    };
    /**
     * Loads the WASM libsvm asynchronously, this is best for browser usage
     */
    SVM.prototype.loadWASM = function () {
        var _this = this;
        return libsvm_2.default
            .load()
            .then(function () {
            if (_this.loaded) {
                throw Errors_1.SVMError('Cannot load an already loaded SVM');
            }
            _this.libsvm = libsvm_2.default;
            _this.initiateAPIs(_this.libsvm);
            _this.loaded = true;
            return _this;
        })
            .catch(function (err) {
            throw Errors_1.WASMError(err);
        });
    };
    /**
     * Trains a model
     * @param args
     */
    SVM.prototype.train = function (args) {
        if (this.deserialized) {
            throw Errors_1.SVMError('Cannot train a deserialized model');
        }
        this.free();
        this.problem = this.createProblem(args);
        var command = this.getCommand({ samples: args.samples });
        this.model = this.train_problem(this.problem, command);
    };
    /**
     * Predict using a single vector of sample
     * @param args
     */
    SVM.prototype.predictOne = function (args) {
        var sample = args.sample;
        if (!this.model) {
            throw Errors_1.SVMError('SVM cannot perform predictOne unless you instantiate it');
        }
        return this.predict_one(this.model, new Uint8Array(new Float64Array(sample).buffer), sample.length);
    };
    /**
     * Predict a matrix
     * @param args
     */
    SVM.prototype.predict = function (args) {
        var samples = args.samples;
        var arr = [];
        for (var i = 0; i < samples.length; i++) {
            arr.push(this.predictOne({ sample: samples[i] }));
        }
        return arr;
    };
    /**
     * Predict the label with probability estimate of many samples.
     * @param args
     */
    SVM.prototype.predictProbability = function (args) {
        var samples = args.samples;
        var arr = [];
        for (var i = 0; i < samples.length; i++) {
            arr.push(this.predictOneProbability({ sample: samples[i] }));
        }
        return arr;
    };
    /**
     * Predict the label with probability estimate.
     * @param args
     */
    SVM.prototype.predictOneProbability = function (args) {
        var sample = args.sample;
        var labels = this.getLabels();
        var nbLabels = labels.length;
        var estimates = this.libsvm._malloc(nbLabels * 8);
        var prediction = this.predict_one_probability(this.model, new Uint8Array(new Float64Array(sample).buffer), sample.length, estimates);
        var estimatesArr = Array.from(this.libsvm.HEAPF64.subarray(estimates / 8, estimates / 8 + nbLabels));
        var result = {
            prediction: prediction,
            estimates: labels.map(function (label, idx) { return ({
                label: label,
                probability: estimatesArr[idx],
            }); }),
        };
        this.libsvm._free(estimates);
        return result;
    };
    /**
     * Predict a regression value with a confidence interval
     * @param args
     */
    SVM.prototype.predictOneInterval = function (args) {
        var sample = args.sample, confidence = args.confidence;
        var interval = this.getInterval({ confidence: confidence });
        var predicted = this.predictOne({ sample: sample });
        return {
            predicted: predicted,
            interval: [predicted - interval, predicted + interval],
        };
    };
    /**
     * Predict regression values with confidence intervals
     * @param args
     */
    SVM.prototype.predictInterval = function (args) {
        var samples = args.samples, confidence = args.confidence;
        var interval = this.getInterval({ confidence: confidence });
        var predicted = this.predict({ samples: samples });
        return predicted.map(function (pred) { return ({
            predicted: pred,
            interval: [pred - interval, pred + interval],
        }); });
    };
    /**
     * Get the array of labels from the model. Useful when creating an SVM instance with SVM.load
     */
    SVM.prototype.getLabels = function () {
        var nLabels = this.svm_get_nr_class(this.model);
        return this.getIntArrayFromModel({ fn: this.svm_get_labels, model: this.model, size: nLabels });
    };
    /**
     * Save the state of the SVM
     */
    SVM.prototype.toJSON = function () {
        return {
            model: this.model,
            options: this.options,
            loaded: this.loaded,
            problem: this.problem,
        };
    };
    /**
     * Load a model from a state
     * @param args
     */
    SVM.prototype.fromJSON = function (args) {
        var model = args.model, problem = args.problem, options = args.options, loaded = args.loaded;
        this.model = model;
        this.options = options;
        this.loaded = loaded;
        this.problem = problem;
    };
    /**
     * Performs k-fold cross-validation (KF-CV). KF-CV separates the data-set into kFold random equally sized partitions,
     * and uses each as a validation set, with all other partitions used in the training set. Observations left over
     * from if kFold does not divide the number of observations are left out of the cross-validation process. If
     * kFold is one, this is equivalent to a leave-on-out cross-validation
     * @param args
     */
    SVM.prototype.crossValidation = function (args) {
        var samples = args.samples, labels = args.labels, kFold = args.kFold;
        if (this.deserialized) {
            throw Errors_1.SVMError('Cannot cross validate on an instance created with SVM.load');
        }
        var problem = this.createProblem({ samples: samples, labels: labels });
        var target = this.libsvm._malloc(labels.length * 8);
        this.svm_cross_validation(problem, this.getCommand({ samples: samples }), kFold, target);
        var data = this.libsvm.HEAPF64.subarray(target / 8, target / 8 + labels.length);
        var arr = Array.from(data);
        this.libsvm._free(target);
        this.free_problem(problem);
        return arr;
    };
    /**
     * Get the indices of the support vectors from the training set passed to the train method.
     */
    SVM.prototype.getSVIndices = function () {
        var nSV = this.svm_get_nr_sv(this.model);
        return this.getIntArrayFromModel({ fn: this.svm_get_sv_indices, model: this.model, size: nSV });
    };
    /**
     * Uses libsvm's serialization method of the model.
     */
    SVM.prototype.serializeModel = function () {
        if (!this.model) {
            throw Errors_1.SVMError('Cannot serialize model. No model was trained');
        }
        var result = this.serialize_model(this.model);
        var str = this.libsvm.Pointer_stringify(result);
        this.libsvm._free(result);
        return str;
    };
    /**
     * Free the memory allocated for the model. Since this memory is stored in the memory model of emscripten, it is
     * allocated within an ArrayBuffer and WILL NOT BE GARBARGE COLLECTED, you have to explicitly free it. So
     * not calling this will result in memory leaks. As of today in the browser, there is no way to hook the
     * garbage collection of the SVM object to free it automatically.
     * Free the memory that was created by the compiled libsvm library to.
     * store the model. This model is reused every time the predict method is called.
     */
    SVM.prototype.free = function () {
        if (this.problem) {
            this.free_problem(this.problem);
            this.problem = null;
        }
        if (this.model !== null) {
            this.svm_free_model(this.model);
            this.model = null;
        }
    };
    // @ts-ignore: Emscripten generated object is used as an input
    SVM.prototype.initiateAPIs = function (libsvm) {
        this.predict_one = libsvm.cwrap('libsvm_predict_one', 'number', ['number', 'array', 'number']);
        this.predict_one_probability = libsvm.cwrap('libsvm_predict_one_probability', 'number', [
            'number',
            'array',
            'number',
            'number',
        ]);
        this.add_instance = libsvm.cwrap('add_instance', null, ['number', 'array', 'number', 'number', 'number']);
        this.create_svm_nodes = libsvm.cwrap('create_svm_nodes', 'number', ['number', 'number']);
        this.train_problem = libsvm.cwrap('libsvm_train_problem', 'number', ['number', 'string']);
        this.svm_get_nr_sv = libsvm.cwrap('svm_get_nr_sv', 'number', ['number']);
        this.svm_get_nr_class = libsvm.cwrap('svm_get_nr_class', 'number', ['number']);
        this.svm_get_sv_indices = libsvm.cwrap('svm_get_sv_indices', null, ['number', 'number']);
        this.svm_get_labels = libsvm.cwrap('svm_get_labels', null, ['number', 'number']);
        this.svm_free_model = libsvm.cwrap('svm_free_model', null, ['number']);
        this.svm_cross_validation = libsvm.cwrap('libsvm_cross_validation', null, ['number', 'string', 'number', 'number']);
        this.svm_get_svr_probability = libsvm.cwrap('svm_get_svr_probability', 'number', ['number']);
        this.free_problem = libsvm.cwrap('free_problem', null, ['number']);
        this.serialize_model = libsvm.cwrap('serialize_model', 'number', ['number']);
        this.deserialize_model = libsvm.cwrap('deserialize_model', 'number', ['string']);
    };
    SVM.prototype.getCommand = function (args) {
        var samples = args.samples;
        var options = {};
        Object.assign(options, this.options, {
            gamma: this.options.gamma ? this.options.gamma : 1 / samples[0].length,
        });
        return Util_1.getCommand(options);
    };
    SVM.prototype.createProblem = function (args) {
        var samples = args.samples, labels = args.labels;
        var nbSamples = samples.length;
        var nbFeatures = labels.length;
        var problem = this.create_svm_nodes(nbSamples, nbFeatures);
        for (var i = 0; i < nbSamples; i++) {
            this.add_instance(problem, new Uint8Array(new Float64Array(samples[i]).buffer), nbFeatures, labels[i], i);
        }
        return problem;
    };
    SVM.prototype.getIntArrayFromModel = function (args) {
        var fn = args.fn, model = args.model, size = args.size;
        var offset = this.libsvm._malloc(size * 4);
        fn(model, offset);
        var data = this.libsvm.HEAP32.subarray(offset / 4, offset / 4 + size);
        // Casting any received from HEAP32 as number[]
        var arr = Array.from(data);
        this.libsvm._free(offset);
        return arr;
    };
    SVM.prototype.getInterval = function (args) {
        var confidence = args.confidence;
        var sigma = this.svm_get_svr_probability(this.model);
        if (sigma === 0) {
            throw Errors_1.SVMError('the model is not a regression with probability estimates');
        }
        if (confidence <= 0 || confidence >= 1) {
            throw Errors_1.SVMError('confidence must be greater than 0 and less than 1');
        }
        var p = (1 - confidence) / 2;
        return sigma * Math.sign(p - 0.5) * Math.log2(1 - 2 * Math.abs(p - 0.5));
    };
    return SVM;
}());
exports.SVM = SVM;
