// machinelearn/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var cluster = __importStar(require("./cluster"));
exports.cluster = cluster;
var datasets = __importStar(require("./datasets"));
exports.datasets = datasets;
var decomposition = __importStar(require("./decomposition"));
exports.decomposition = decomposition;
var ensemble = __importStar(require("./ensemble"));
exports.ensemble = ensemble;
var feature_extraction = __importStar(require("./feature_extraction"));
exports.feature_extraction = feature_extraction;
var linear_model = __importStar(require("./linear_model"));
exports.linear_model = linear_model;
var metrics = __importStar(require("./metrics"));
exports.metrics = metrics;
var model_selection = __importStar(require("./model_selection"));
exports.model_selection = model_selection;
var neighbors = __importStar(require("./neighbors"));
exports.neighbors = neighbors;
var preprocessing = __importStar(require("./preprocessing"));
exports.preprocessing = preprocessing;
var svm = __importStar(require("./svm"));
exports.svm = svm;
var tree = __importStar(require("./tree"));
exports.tree = tree;
