// machinelearn/utils/Errors.js
"use strict";
// NOTE: Below custom errors are hack because Jest has a bug with asserting error types
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The error is used for class initiation failures due to invalid arguments.
 * @ignore
 */
exports.ConstructionError = function (message) {
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.message = message;
};
/**
 * The error is used for any validation errors. Such as an argument type check failure would raise this error.
 * @ignore
 */
exports.ValidationError = function (message) {
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.message = message;
};
/**
 * @ignore
 */
exports.Validation1DMatrixError = function (message) {
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.message = message;
};
/**
 * @ignore
 */
exports.Validation2DMatrixError = function (message) {
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.message = message;
};
/**
 * The error is used when a matrix does not contain a consistent type for its elements
 * @ignore
 */
exports.ValidationMatrixTypeError = function (message) {
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.message = message;
};
/**
 * @ignore
 */
exports.ValidationClassMismatch = function (message) {
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.message = message;
};
/**
 * @ignore
 */
exports.ValidationKeyNotFoundError = function (message) {
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.message = message;
};
/**
 * @ignore
 */
exports.ValidationInconsistentShape = function (message) {
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.message = message;
};
