// machinelearn/utils/validation.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var tf = __importStar(require("@tensorflow/tfjs"));
var _ = __importStar(require("lodash"));
var util_1 = require("util");
var Errors_1 = require("./Errors");
var tensors_1 = require("./tensors");
/**
 * Check below array conditions
 * - multiclass
 *    - e.g. [ [1, 2], [2, 3] ]
 *      Then it sets multiclass value to true
 * - isArray<boolean>
 *   If the given arr is an array then the value is true else false
 * @param arr
 * @returns {any}
 * @ignore
 */
function checkArray(arr) {
    var result = {
        isArray: false,
        multiclass: false,
    };
    // Setting isArray flag
    if (_.isArray(arr)) {
        result = _.set(result, 'isArray', true);
    }
    else {
        result = _.set(result, 'isArray', false);
    }
    // Setting multiclass flag
    var firstElm = _.get(arr, '[0]');
    if (_.isArray(firstElm)) {
        result = _.set(result, 'multiclass', true);
    }
    else {
        result = _.set(result, 'multiclass', false);
    }
    return result;
}
exports.checkArray = checkArray;
/**
 * Validates the input matrix's types with the targetted types.
 * Specified target types must be one of https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof#Description
 *
 * @example
 * validateMatrixType([['z', 'z']],['string']); // no errors
 * validateMatrixType([['z', 'z']],['test']); // error: Input matrix type of ["string"] does not match with the target types ["test"]
 *
 * @param X - The input matrix
 * @param targetTypes - Target matrix types
 * @ignore
 */
function validateMatrixType(X, targetTypes) {
    var flatX = _.flattenDeep(X);
    var xTypes = _.uniq(flatX.map(function (x) { return typeof x; }));
    var sortedXTypes = _.sortBy(xTypes, function (x) { return x; });
    var sortedTargetTypes = _.sortBy(targetTypes, function (x) { return x; });
    if (!_.isEqual(sortedXTypes, sortedTargetTypes)) {
        throw new Errors_1.ValidationMatrixTypeError("Input matrix type of " + JSON.stringify(sortedXTypes) + " does not match with the target types " + JSON.stringify(sortedTargetTypes));
    }
}
exports.validateMatrixType = validateMatrixType;
/**
 * Check that X and y have the same size across the first axis
 *
 * @example
 * validateTrainInputs([ [1, 2], [3, 4] ], [ 1, 2 ]) // No errors
 * validateTrainInputs([ [[1, 2], [3, 3]], [[1, 2], [3, 3]] ], [ 1, 2 ]) // Error: The matrix is not 1D shaped: [ [[1, 2], [3, 3]], [[1, 2], [3, 3]] ] of [2, 2, 2]
 *
 * @param X
 * @param y
 * @ignore
 */
function validateFitInputs(X, y) {
    if (!Array.isArray(X)) {
        throw new Errors_1.ValidationError('validateFitInputs received a non-array input X');
    }
    if (!Array.isArray(y)) {
        throw new Errors_1.ValidationError('validateFitInputs received a non-array input y');
    }
    // Check X is always a matrix
    var sampleShape = tensors_1.inferShape(X);
    // Check y is always a vector
    var targetShape = tensors_1.inferShape(y);
    if (sampleShape[0] !== targetShape[0]) {
        throw new Errors_1.ValidationClassMismatch("Number of labels=" + targetShape[0] + " does not math number of samples=" + sampleShape[0]);
    }
}
exports.validateFitInputs = validateFitInputs;
/**
 * Validate the matrix is 1D shaped by checking the shape's length is exactly  1
 * @param X
 * @ignore
 */
function validateMatrix1D(X) {
    if (!util_1.isArray(X)) {
        throw new Errors_1.ValidationError('validateMatrix1D has received a non-array argument');
    }
    var shape = tensors_1.inferShape(X);
    if (shape.length !== 1 || shape[0] === 0) {
        throw new Errors_1.Validation1DMatrixError("The matrix is not 1D shaped: " + JSON.stringify(X) + " of " + JSON.stringify(shape));
    }
    return X;
}
exports.validateMatrix1D = validateMatrix1D;
/**
 * Validate the matrix is 2D shaped by checking the shape's length is exactly 2
 * @param X - An input array
 * @ignore
 */
function validateMatrix2D(X) {
    if (!Array.isArray(X)) {
        throw new Errors_1.ValidationError('validateMatrix2D has received a non-array argument');
    }
    var shape = tensors_1.inferShape(X);
    if (shape.length !== 2) {
        throw new Errors_1.Validation2DMatrixError("The matrix is not 2D shaped: " + JSON.stringify(X) + " of " + JSON.stringify(shape));
    }
    return X;
}
exports.validateMatrix2D = validateMatrix2D;
/**
 * Checks that provided X matrix has the same number of features as model matrix
 * @param X - matrix to check
 * @param reference - reference matrix
 * @throws ValidationError - in case number of features doesn't match
 * @ignore
 */
exports.validateFeaturesConsistency = function (X, reference) {
    var xShape = tensors_1.inferShape(X);
    var referenceShape = tensors_1.inferShape(reference);
    var xNumFeatures = xShape.length === 1 ? 1 : xShape[1];
    var referenceNumFeatures = referenceShape[0];
    if (xNumFeatures !== referenceNumFeatures) {
        throw new Errors_1.ValidationError("Provided X has incorrect number of features. Should have: " + referenceNumFeatures + ", got: " + xNumFeatures);
    }
};
/**
 * Checks that provided X matrix has the same number of features as model matrix
 * @param y_true - matrix to check
 * @param y_pred - matrix to check
 * @throws ValidationError - in case any of the params are empty
 * @throws ValidationError - in case y_true and y_pred are of different shape
 * @ignore
 */
function validateShapesEqual(y_true, y_pred) {
    if (y_true === void 0) { y_true = null; }
    if (y_pred === void 0) { y_pred = null; }
    var yTrueTensor = tf.tensor(y_true);
    var yPredTensor = tf.tensor(y_pred);
    var yTrueShape = yTrueTensor.shape;
    var yPredShape = yPredTensor.shape;
    // Validation 1: empty array check
    if (yTrueShape[0] === 0 || yPredShape[0] === 0) {
        throw new Errors_1.ValidationError("y_true " + JSON.stringify(y_true) + " and y_pred " + JSON.stringify(y_pred) + " cannot be empty");
    }
    // Validation 2: Same shape
    if (!_.isEqual(yTrueShape, yPredShape)) {
        throw new Errors_1.ValidationError("Shapes of y_true " + JSON.stringify(yTrueShape) + " and y_pred " + JSON.stringify(yPredShape) + " should be equal");
    }
    return [yTrueTensor, yPredTensor];
}
exports.validateShapesEqual = validateShapesEqual;
