// machinelearn/neighbors/classification.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var MathExtra_1 = __importDefault(require("../utils/MathExtra"));
var tensors_1 = require("../utils/tensors");
var validation_1 = require("../utils/validation");
var KDTree_1 = __importDefault(require("./KDTree"));
var euclideanDistance = MathExtra_1.default.euclideanDistance, manhattanDistance = MathExtra_1.default.manhattanDistance;
var DIST_EUC = 'euclidean';
var DIST_MAN = 'manhattan';
var TYPE_KD = 'kdtree';
/**
 * Classifier implementing the k-nearest neighbors vote.
 *
 * @example
 * const knn = new KNeighborsClassifier();
 * const X = [[0, 0, 0], [0, 1, 1], [1, 1, 0], [2, 2, 2], [1, 2, 2], [2, 1, 2]];
 * const y = [0, 0, 0, 1, 1, 1];
 * knn.fit(X ,y);
 * console.log(knn.predict([1, 2])); // predicts 1
 */
var KNeighborsClassifier = /** @class */ (function () {
    /**
     * @param {string} distance - Choice of distance function, should choose between euclidean | manhattan
     * @param {number} k - Number of neighbors to classify
     * @param {string} type - Type of algorithm to use, choose between kdtree(default) | balltree | simple
     */
    function KNeighborsClassifier(_a) {
        var _b = _a === void 0 ? {
            // Default value on empty constructor
            distance: DIST_EUC,
            k: 0,
            type: TYPE_KD,
        } : _a, 
        // Each object param default value
        _c = _b.distance, 
        // Each object param default value
        distance = _c === void 0 ? DIST_EUC : _c, _d = _b.k, k = _d === void 0 ? 0 : _d, _e = _b.type, type = _e === void 0 ? TYPE_KD : _e;
        this.type = null;
        this.tree = null;
        this.k = null;
        this.classes = null;
        this.distance = null;
        var options = {
            distance: distance,
            k: k,
            type: type,
        };
        // Handling distance
        if (options.distance === DIST_EUC) {
            this.distance = euclideanDistance;
        }
        else if (options.distance === DIST_MAN) {
            this.distance = manhattanDistance;
        }
        else {
            throw new Error("Unrecognised type of distance " + options.distance + " was received");
        }
        this.k = options.k;
        this.type = options.type;
    }
    /**
     * Train the classifier with input and output data
     * @param {any} X - Training data.
     * @param {any} y - Target data.
     */
    KNeighborsClassifier.prototype.fit = function (X, y) {
        validation_1.validateFitInputs(X, y);
        // Getting the classes from y
        var classes = lodash_1.uniqBy(y, function (c) { return c; });
        // Setting k; if it's null, use the class length
        var k = this.k ? this.k : classes.length + 1;
        //  Constructing the points placeholder
        var points = new Array(X.length);
        for (var i = 0; i < points.length; ++i) {
            points[i] = X[i].slice();
        }
        for (var i = 0; i < y.length; ++i) {
            points[i].push(y[i]);
        }
        // Building a tree or algo according to this.type
        if (this.type === TYPE_KD) {
            this.tree = new KDTree_1.default(points, this.distance);
        }
        this.k = k;
        this.classes = classes;
    };
    /**
     * Return the model's state as a JSON object
     * @return {object} JSON KNN model.
     */
    KNeighborsClassifier.prototype.toJSON = function () {
        return {
            classes: this.classes,
            distance: this.distance,
            k: this.k,
            tree: this.tree,
            type: this.type,
        };
    };
    /**
     * Restores the model from a JSON checkpoint
     * @param {any} classes
     * @param {any} distance
     * @param {any} k
     * @param {any} tree
     * @param {any} type
     */
    KNeighborsClassifier.prototype.fromJSON = function (_a) {
        var _b = _a.classes, classes = _b === void 0 ? null : _b, _c = _a.distance, distance = _c === void 0 ? null : _c, _d = _a.k, k = _d === void 0 ? null : _d, _e = _a.tree, tree = _e === void 0 ? null : _e, _f = _a.type, type = _f === void 0 ? null : _f;
        if (!classes || !distance || !k || !tree || !type) {
            throw new Error('You must provide classes, distance, k, tree and type to restore the KNearestNeighbor');
        }
        this.classes = classes;
        this.distance = distance;
        this.k = k;
        this.tree = tree;
        this.type = type;
    };
    /**
     * Predict single value from a list of data
     * @param {Array} X - Prediction data.
     * @returns number
     */
    KNeighborsClassifier.prototype.predict = function (X) {
        var _this = this;
        var shape = tensors_1.inferShape(X);
        if (shape.length === 1) {
            return this.getSinglePred(X);
        }
        else if (shape.length === 2) {
            return lodash_1.map(X, function (currentItem) { return _this.getSinglePred(currentItem); });
        }
        else {
            throw new TypeError('The dataset is neither an array or a matrix');
        }
    };
    /**
     * Runs a single prediction against an array based on kdTree or balltree or
     * simple algo
     * @param array
     * @returns {{}}
     */
    KNeighborsClassifier.prototype.getSinglePred = function (array) {
        if (this.tree) {
            return this.getTreeBasedPrediction(array);
        }
        else {
            // Run the simple KNN algorithm
            return 0;
        }
    };
    /**
     * Get the class with the max point
     * @param current
     * @returns {{}}
     * @ignore
     */
    KNeighborsClassifier.prototype.getTreeBasedPrediction = function (current) {
        var nearestPoints = this.tree.nearest(current, this.k);
        var pointsPerClass = {};
        var predictedClass = -1;
        var maxPoints = -1;
        var lastElement = nearestPoints[0][0].length - 1;
        // Initialising the points placeholder per class
        for (var j = 0; j < this.classes.length; j++) {
            pointsPerClass[this.classes[j]] = 0;
        }
        // Voting the max value
        for (var i = 0; i < nearestPoints.length; ++i) {
            var currentClass = nearestPoints[i][0][lastElement];
            var currentPoints = ++pointsPerClass[currentClass];
            if (currentPoints > maxPoints) {
                predictedClass = currentClass;
                maxPoints = currentPoints;
            }
        }
        return predictedClass;
    };
    return KNeighborsClassifier;
}());
exports.KNeighborsClassifier = KNeighborsClassifier;
