// machinelearn/cluster/k_means.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var _ = __importStar(require("lodash"));
var Random = __importStar(require("random-js"));
var MathExtra_1 = __importDefault(require("../utils/MathExtra"));
var validation_1 = require("../utils/validation");
/**
 * K-Means clustering
 *
 * @example
 * import { KMeans } from 'machinelearn/cluster';
 *
 * const kmean = new KMeans({ k: 2 });
 * const clusters = kmean.fit([[1, 2], [1, 4], [1, 0], [4, 2], [4, 4], [4, 0]]);
 *
 * const result = kmean.predict([[0, 0], [4, 4]]);
 * // results in: [0, 1]
 */
var KMeans = /** @class */ (function () {
    /**
     *
     * @param distance - Choice of distance method. Defaulting to euclidean
     * @param k - Number of clusters
     * @param maxIteration - Relative tolerance with regards to inertia to declare convergence
     * @param randomState - Random state value for sorting centroids during the getInitialCentroid phase
     */
    function KMeans(_a) {
        var _b = _a === void 0 ? {
            distance: 'euclidean',
            k: 3,
            maxIteration: 300,
            randomState: 0,
        } : _a, _c = _b.distance, distance = _c === void 0 ? 'euclidean' : _c, _d = _b.k, k = _d === void 0 ? 3 : _d, _e = _b.maxIteration, maxIteration = _e === void 0 ? 300 : _e, _f = _b.randomState, randomState = _f === void 0 ? 0 : _f;
        this.k = k;
        // Assigning a distance method
        var distanceType = distance;
        switch (distanceType) {
            case 'euclidean':
                this.distance = MathExtra_1.default.euclideanDistance;
                break;
            case 'manhattan':
                this.distance = MathExtra_1.default.manhattanDistance;
                break;
            default:
                throw new Error("Unknown distance type " + distanceType);
        }
        this.randomState = randomState;
        this.maxIteration = maxIteration;
        this.centroids = [];
    }
    /**
     * Compute k-means clustering.
     * @param {any} X - array-like or sparse matrix of shape = [n_samples, n_features]
     * @returns {{centroids: number[]; clusters: number[]}}
     */
    KMeans.prototype.fit = function (X) {
        var _this = this;
        if (X === void 0) { X = null; }
        validation_1.validateMatrix2D(X);
        this.assignment = new Array(_.size(X));
        this.centroids = this.getInitialCentroids(X, this.k);
        this.clusters = new Array(this.k);
        // Flag to check the convergence
        var movement = true;
        // Looping only within the maxIteration boundary
        for (var iter = 0; iter < this.maxIteration && movement; iter++) {
            // find the distance between the point and cluster; choose the nearest centroid
            _.forEach(X, function (data, i) {
                _this.assignment[i] = _this.getClosestCentroids(data, _this.centroids, _this.distance);
            });
            // Flag set to false; giving opportunity to stop the loop upon the covergence
            movement = false;
            // Updating the location of each centroid
            for (var j = 0; j < this.k; j++) {
                var assigned = [];
                for (var i = 0; i < this.assignment.length; i++) {
                    if (this.assignment[i] === j) {
                        assigned.push(X[i]);
                    }
                }
                if (!assigned.length) {
                    continue;
                }
                // Getting the original data point
                // TODO: Fix any type
                var centroid = this.centroids[j];
                var newCentroid = new Array(centroid.length);
                for (var g = 0; g < centroid.length; g++) {
                    var sum = 0;
                    for (var i = 0; i < assigned.length; i++) {
                        sum += assigned[i][g];
                    }
                    newCentroid[g] = sum / assigned.length;
                    // Does not converge yet
                    if (newCentroid[g] !== centroid[g]) {
                        movement = true;
                    }
                }
                this.centroids[j] = newCentroid;
                this.clusters[j] = assigned;
            }
        }
    };
    /**
     * Predicts the cluster index with the given X
     * @param {any} X - array-like or sparse matrix of shape = [n_samples, n_features]
     * @returns {number[]}
     */
    KMeans.prototype.predict = function (X) {
        var _this = this;
        if (X === void 0) { X = null; }
        validation_1.validateMatrix2D(X);
        return _.map(X, function (data) {
            return _this.getClosestCentroids(data, _this.centroids, _this.distance);
        });
    };
    /**
     * Get the model details in JSON format
     * @returns {{k: number; clusters: number[]; centroids: number[]}}
     */
    KMeans.prototype.toJSON = function () {
        return {
            centroids: this.centroids,
            clusters: this.clusters,
            k: this.k,
        };
    };
    /**
     * Restores the model from checkpoints
     * @param {number} k
     * @param {number[]} clusters
     * @param {number[]} centroids
     */
    KMeans.prototype.fromJSON = function (_a) {
        var _b = _a.k, k = _b === void 0 ? null : _b, _c = _a.clusters, clusters = _c === void 0 ? null : _c, _d = _a.centroids, centroids = _d === void 0 ? null : _d;
        if (!k || !clusters || !centroids) {
            throw new Error('You must provide all the parameters include k, clusters and centroids');
        }
        this.k = k;
        this.clusters = clusters;
        this.centroids = centroids;
    };
    /**
     * Get initial centroids from X of k
     * @param {number[]} X
     * @param {number} k
     * @returns {number[]}
     */
    KMeans.prototype.getInitialCentroids = function (X, k) {
        // Create an initial copy
        var centroids = _.clone(X);
        // Sort the centroid randomly if the randomState is greater than 0
        if (this.randomState > 0) {
            var randomEngine_1 = Random.engines.mt19937();
            randomEngine_1.seed(this.randomState);
            centroids.sort(function () {
                var randomInt = Random.integer(0, 1)(randomEngine_1);
                return Math.round(randomInt) - 0.5;
            });
        }
        return centroids.slice(0, k);
    };
    /**
     * Get closest centroids based on the passed in distance method
     * @param {number[]} data
     * @param {number[]} centroids
     * @param distance
     * @returns {number}
     */
    KMeans.prototype.getClosestCentroids = function (data, centroids, distance) {
        var min = Infinity;
        var index = 0;
        _.forEach(centroids, function (centroid, i) {
            var dist = distance(data, centroid);
            if (dist < min) {
                min = dist;
                index = i;
            }
        });
        return index;
    };
    return KMeans;
}());
exports.KMeans = KMeans;
