// machinelearn/linear_model/logistic_regression.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var tf = __importStar(require("@tensorflow/tfjs"));
var tensors_1 = require("../utils/tensors");
var validation_1 = require("../utils/validation");
/**
 * Logistic Regression (aka logit, MaxEnt) classifier.
 *
 *
 * Logistic regression is named for the function used at the core of the method, the logistic function.
 * The logistic function, also called the sigmoid function was developed by statisticians to describe properties of
 * population growth in ecology, rising quickly and maxing out at the carrying capacity of the environment.
 * It’s an S-shaped curve that can take any real-valued number and map it into a value between 0 and 1,
 * but never exactly at those limits.
 *
 * 1 / (1 + e^-value)
 *
 * @example
 * import { LogisticRegression } from 'machinelearn/linear_model';
 * import { HeartDisease } from 'machinelearn/datasets';
 *
 * (async function() {
 *   const { data, targets } = await heartDisease.load();
 *   const { xTest, xTrain, yTest } = train_test_split(data, targets);
 *
 *   const lr = new LogisticRegression();
 *   lr.fit(xTrain, yTrain);
 *
 *   lr.predict(yTest);
 * });
 *
 */
var LogisticRegression = /** @class */ (function () {
    /**
     * @param learning_rate - Model learning rate
     * @param num_iterations - Number of iterations to run gradient descent fo
     */
    function LogisticRegression(_a) {
        var _b = _a === void 0 ? {
            learning_rate: 0.001,
            num_iterations: 4000,
        } : _a, _c = _b.learning_rate, learning_rate = _c === void 0 ? 0.001 : _c, _d = _b.num_iterations, num_iterations = _d === void 0 ? 4000 : _d;
        this.learningRate = learning_rate;
        this.numIterations = num_iterations;
    }
    /**
     * Fit the model according to the given training data.
     * @param X - A matrix of samples
     * @param y - A matrix of targets
     */
    LogisticRegression.prototype.fit = function (X, y) {
        if (X === void 0) { X = null; }
        if (y === void 0) { y = null; }
        var xWrapped = tensors_1.ensure2DMatrix(X);
        validation_1.validateFitInputs(xWrapped, y);
        this.initWeights(xWrapped);
        var tensorX = tf.tensor2d(xWrapped);
        var tensorY = tf.tensor1d(y);
        for (var i = 0; i < this.numIterations; ++i) {
            var predictions = tf.sigmoid(tensorX.dot(this.weights));
            var gradient = tf.mul(tensorY.sub(predictions).dot(tensorX), -1);
            this.weights = this.weights.sub(tf.mul(this.learningRate, gradient));
        }
    };
    /**
     * Predict class labels for samples in X.
     * @param X - A matrix of test data
     * @returns An array of predicted classes
     */
    LogisticRegression.prototype.predict = function (X) {
        if (X === void 0) { X = null; }
        validation_1.validateFeaturesConsistency(X, this.weights.arraySync());
        var xWrapped = tensors_1.ensure2DMatrix(X);
        var syncResult = tf.round(tf.sigmoid(tf.tensor2d(xWrapped).dot(this.weights))).arraySync();
        return validation_1.validateMatrix1D(syncResult);
    };
    /**
     * Get the model details in JSON format
     */
    LogisticRegression.prototype.toJSON = function () {
        return {
            weights: this.weights.arraySync(),
            learning_rate: this.learningRate,
        };
    };
    /**
     * Restore the model from a checkpoint
     */
    LogisticRegression.prototype.fromJSON = function (_a) {
        var _b = _a === void 0 ? {
            weights: null,
            learning_rate: 0.001,
        } : _a, 
        /**
         * Model training weights
         */
        _c = _b.weights, 
        /**
         * Model training weights
         */
        weights = _c === void 0 ? null : _c, 
        /**
         * Model learning rate
         */
        _d = _b.learning_rate, 
        /**
         * Model learning rate
         */
        learning_rate = _d === void 0 ? null : _d;
        this.weights = tf.tensor1d(weights);
        this.learningRate = learning_rate;
    };
    LogisticRegression.prototype.initWeights = function (X) {
        var shape = tensors_1.inferShape(X);
        var numFeatures = shape[1];
        var limit = 1 / Math.sqrt(numFeatures);
        this.weights = tf.randomUniform([numFeatures], -limit, limit);
    };
    return LogisticRegression;
}());
exports.LogisticRegression = LogisticRegression;
