// machinelearn/linear_model/stochastic_gradient.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var tf = __importStar(require("@tensorflow/tfjs"));
var lodash_1 = require("lodash");
var Random = __importStar(require("random-js"));
var validation_1 = require("../utils/validation");
var TypeLoss;
(function (TypeLoss) {
    TypeLoss["L1"] = "L1";
    TypeLoss["L2"] = "L2";
    TypeLoss["L1L2"] = "L1L2";
})(TypeLoss = exports.TypeLoss || (exports.TypeLoss = {}));
/**
 * Ordinary base class for SGD classier or regressor
 * @ignore
 */
var BaseSGD = /** @class */ (function () {
    /**
     * @param preprocess - preprocess methodology can be either minmax or null. Default is minmax.
     * @param learning_rate - Used to limit the amount each coefficient is corrected each time it is updated.
     * @param epochs - Number of iterations.
     * @param clone - To clone the passed in dataset.
     */
    function BaseSGD(_a) {
        var _b = _a === void 0 ? {
            learning_rate: 0.0001,
            epochs: 10000,
            clone: true,
            random_state: null,
            loss: TypeLoss.L2,
            reg_factor: null,
        } : _a, _c = _b.learning_rate, learning_rate = _c === void 0 ? 0.0001 : _c, _d = _b.epochs, epochs = _d === void 0 ? 10000 : _d, _e = _b.clone, clone = _e === void 0 ? true : _e, _f = _b.random_state, random_state = _f === void 0 ? null : _f, _g = _b.loss, loss = _g === void 0 ? TypeLoss.L2 : _g, _h = _b.reg_factor, reg_factor = _h === void 0 ? null : _h;
        this.clone = true;
        this.weights = null;
        this.learningRate = learning_rate;
        this.epochs = epochs;
        this.clone = clone;
        this.randomState = random_state;
        this.loss = loss;
        this.regFactor = reg_factor;
        // Setting a loss function according to the input option
        if (this.loss === TypeLoss.L1 && this.regFactor) {
            this.loss = tf.regularizers.l1({
                l1: this.regFactor.l1,
            });
        }
        else if (this.loss === TypeLoss.L1L2 && this.regFactor) {
            this.loss = tf.regularizers.l1l2({
                l1: this.regFactor.l1,
                l2: this.regFactor.l2,
            });
        }
        else if (this.loss === TypeLoss.L2 && this.regFactor) {
            this.loss = tf.regularizers.l2({
                l2: this.regFactor.l2,
            });
        }
        else {
            this.loss = tf.regularizers.l2();
        }
        // Random Engine
        if (Number.isInteger(this.randomState)) {
            this.randomEngine = Random.engines.mt19937().seed(this.randomState);
        }
        else {
            this.randomEngine = Random.engines.mt19937().autoSeed();
        }
    }
    /**
     * Train the base SGD
     * @param X - Matrix of data
     * @param y - Matrix of targets
     */
    BaseSGD.prototype.fit = function (X, y) {
        if (X === void 0) { X = null; }
        if (y === void 0) { y = null; }
        validation_1.validateFitInputs(X, y);
        // holds all the preprocessed X values
        // Clone according to the clone flag
        var clonedX = this.clone ? lodash_1.cloneDeep(X) : X;
        var clonedY = this.clone ? lodash_1.cloneDeep(y) : y;
        this.sgd(clonedX, clonedY);
    };
    /**
     * Save the model's checkpoint
     */
    BaseSGD.prototype.toJSON = function () {
        return {
            learning_rate: this.learningRate,
            epochs: this.epochs,
            weights: __spread(this.weights.dataSync()),
            random_state: this.randomState,
        };
    };
    /**
     * Restore the model from a checkpoint
     * @param learning_rate - Training learning rate
     * @param epochs - Number of model's training epochs
     * @param weights - Model's training state
     * @param random_state - Static random state for the model initialization
     */
    BaseSGD.prototype.fromJSON = function (_a) {
        var _b = _a === void 0 ? {
            learning_rate: 0.0001,
            epochs: 10000,
            weights: [],
            random_state: null,
        } : _a, _c = _b.learning_rate, learning_rate = _c === void 0 ? 0.0001 : _c, _d = _b.epochs, epochs = _d === void 0 ? 10000 : _d, _e = _b.weights, weights = _e === void 0 ? [] : _e, _f = _b.random_state, random_state = _f === void 0 ? null : _f;
        this.learningRate = learning_rate;
        this.epochs = epochs;
        this.weights = tf.tensor(weights);
        this.randomState = random_state;
    };
    /**
     * Predictions according to the passed in test set
     * @param X - Matrix of data
     */
    BaseSGD.prototype.predict = function (X) {
        if (X === void 0) { X = null; }
        validation_1.validateMatrix2D(X);
        // Adding bias
        var biasX = this.addBias(X);
        var tensorX = tf.tensor(biasX);
        var yPred = tensorX.dot(this.weights);
        return __spread(yPred.dataSync());
    };
    /**
     * Initialize weights based on the number of features
     *
     * @example
     * initializeWeights(3);
     * // this.w = [-0.213981293, 0.12938219, 0.34875439]
     *
     * @param nFeatures
     */
    BaseSGD.prototype.initializeWeights = function (nFeatures) {
        var _this = this;
        var limit = 1 / Math.sqrt(nFeatures);
        var distribution = Random.real(-limit, limit);
        var getRand = function () { return distribution(_this.randomEngine); };
        this.weights = tf.tensor1d(lodash_1.range(0, nFeatures).map(function () { return getRand(); }));
    };
    /**
     * Adding bias to a given array
     *
     * @example
     * addBias([[1, 2], [3, 4]], 1);
     * // [[1, 1, 2], [1, 3, 4]]
     *
     * @param X
     * @param bias
     */
    BaseSGD.prototype.addBias = function (X, bias) {
        if (bias === void 0) { bias = 1; }
        // TODO: Is there a TF way to achieve it?
        return X.reduce(function (sum, cur) {
            sum.push([bias].concat(cur));
            return sum;
        }, []);
    };
    /**
     * SGD based on linear model to calculate coefficient
     * @param X - training data
     * @param y - target data
     */
    BaseSGD.prototype.sgd = function (X, y) {
        var tensorX = tf.tensor2d(this.addBias(X));
        this.initializeWeights(tensorX.shape[1]);
        var tensorY = tf.tensor1d(y);
        var tensorLR = tf.tensor(this.learningRate);
        for (var e = 0; e < this.epochs; e++) {
            var yPred = tensorX.dot(this.weights);
            var gradW = tensorY
                .sub(yPred)
                .neg()
                .dot(tensorX)
                .add(this.loss.apply(this.weights));
            this.weights = this.weights.sub(tensorLR.mul(gradW));
        }
    };
    return BaseSGD;
}());
exports.BaseSGD = BaseSGD;
/**
 * Linear classifiers (SVM, logistic regression, a.o.) with SGD training.
 *
 * This estimator implements regularized linear models with
 * stochastic gradient descent (SGD) learning: the gradient of
 * the loss is estimated each sample at a time and the model is
 * updated along the way with a decreasing strength schedule
 * (aka learning rate). SGD allows minibatch (online/out-of-core)
 * learning, see the partial_fit method. For best results using
 * the default learning rate schedule, the data should have zero mean
 * and unit variance.
 *
 * @example
 * import { SGDClassifier } from 'machinelearn/linear_model';
 * const clf = new SGDClassifier();
 * const X = [[0., 0.], [1., 1.]];
 * const y = [0, 1];
 * clf.fit(X ,y);
 * clf.predict([[2., 2.]]); // result: [ 1 ]
 *
 */
var SGDClassifier = /** @class */ (function (_super) {
    __extends(SGDClassifier, _super);
    function SGDClassifier() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Predicted values with Math.round applied
     * @param X - Matrix of data
     */
    SGDClassifier.prototype.predict = function (X) {
        if (X === void 0) { X = null; }
        var results = _super.prototype.predict.call(this, X);
        return results.map(function (x) { return Math.round(x); });
    };
    return SGDClassifier;
}(BaseSGD));
exports.SGDClassifier = SGDClassifier;
/**
 * Linear model fitted by minimizing a regularized empirical loss with SGD
 * SGD stands for Stochastic Gradient Descent: the gradient of the loss
 * is estimated each sample at a time and the model is updated along
 * the way with a decreasing strength schedule (aka learning rate).
 *
 * @example
 * import { SGDRegressor } from 'machinelearn/linear_model';
 * const reg = new SGDRegressor();
 * const X = [[0., 0.], [1., 1.]];
 * const y = [0, 1];
 * reg.fit(X, y);
 * reg.predict([[2., 2.]]); // result: [ 1.281828588248001 ]
 *
 */
var SGDRegressor = /** @class */ (function (_super) {
    __extends(SGDRegressor, _super);
    function SGDRegressor() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Predicted values
     * @param X - Matrix of data
     */
    SGDRegressor.prototype.predict = function (X) {
        if (X === void 0) { X = null; }
        return _super.prototype.predict.call(this, X);
    };
    return SGDRegressor;
}(BaseSGD));
exports.SGDRegressor = SGDRegressor;
