// machinelearn/tree/tree.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var Random = __importStar(require("random-js"));
var validation_1 = require("../utils/validation");
/**
 * Question used by decision tree algorithm to determine whether to split branch or not
 * @ignore
 */
var Question = /** @class */ (function () {
    function Question(features, column, value) {
        if (features === void 0) { features = null; }
        this.features = [];
        this.column = null;
        this.value = null;
        this.features = features;
        this.column = column;
        this.value = value;
    }
    Question.prototype.match = function (example) {
        var val = example[this.column];
        if (typeof val === 'number') {
            return val >= this.value;
        }
        else {
            return val === this.value;
        }
    };
    Question.prototype.toString = function () {
        if (!this.features) {
            throw Error('You must provide feature labels in order to render toString!');
        }
        var condition = typeof this.value === 'number' ? '>=' : '==';
        return "Is " + this.features[this.column] + " " + condition + " " + this.value;
    };
    return Question;
}());
exports.Question = Question;
/**
 * According to the given targets array, count occurrences into an object.
 * @param {any[]} targets - list of class: count
 * @returns {}
 * @ignore
 */
function classCounts(targets) {
    var result = {};
    for (var i = 0; i < targets.length; i++) {
        var target = targets[i];
        var count = result[target]; // the current
        if (typeof count === 'number' && count > 0) {
            result[target] = {
                value: target,
                count: count + 1,
            };
        }
        else {
            result[target] = {
                value: target,
                count: 1,
            };
        }
    }
    return result;
}
exports.classCounts = classCounts;
/**
 * A leaf node that classifies data.
 * @ignore
 */
var Leaf = /** @class */ (function () {
    function Leaf(y) {
        this.prediction = null;
        var counts = classCounts(y);
        var keys = Object.keys(counts); // Retrieving the keys for looping
        // Variable holders
        var maxCount = 0;
        var maxValue = null;
        // Finding the max count key(actual prediction value)
        for (var i = 0; i < keys.length; i++) {
            var key = keys[i];
            var count = counts[key].count;
            var value = counts[key].value;
            if (count > maxCount) {
                maxValue = value;
                maxCount = count;
            }
        }
        this.prediction = maxValue;
    }
    return Leaf;
}());
exports.Leaf = Leaf;
/**
 * It holds a reference to the question, and to the two children nodes
 * @ignore
 */
var DecisionNode = /** @class */ (function () {
    function DecisionNode(question, trueBranch, falseBranch) {
        this.question = null;
        this.trueBranch = null;
        this.falseBranch = null;
        this.question = question;
        this.trueBranch = trueBranch;
        this.falseBranch = falseBranch;
    }
    return DecisionNode;
}());
exports.DecisionNode = DecisionNode;
/**
 * A decision tree classifier.
 *
 * @example
 * import { DecisionTreeClassifier } from 'machinelearn/tree';
 * const features = ['color', 'diameter', 'label'];
 * const decision = new DecisionTreeClassifier({ featureLabels: features });
 *
 * const X = [['Green', 3], ['Yellow', 3], ['Red', 1], ['Red', 1], ['Yellow', 3]];
 * const y = ['Apple', 'Apple', 'Grape', 'Grape', 'Lemon'];
 * decision.fit({ X, y });
 * decision.printTree(); // try it out yourself! =)
 *
 * decision.predict({ X: [['Green', 3]] }); // [ 'Apple' ]
 * decision.predict({ X }); // [ [ 'Apple' ], [ 'Apple', 'Lemon' ], [ 'Grape', 'Grape' ], [ 'Grape', 'Grape' ], [ 'Apple', 'Lemon' ] ]
 *
 * @example
 * import { DecisionTreeClassifier } from 'machinelearn/tree';
 * const decision = new DecisionTreeClassifier({ featureLabels: null });
 *
 * const X = [[0, 0], [1, 1]];
 * const Y = [0, 1];
 * decision.fit({ X, y });
 * decision2.predict({ row: [[2, 2]] }); // [ 1 ]
 */
var DecisionTreeClassifier = /** @class */ (function () {
    /**
     *
     * @param featureLabels - Literal names for each feature to be used while printing the tree out as a string
     * @param verbose - Logs the progress of the tree construction as console.info
     * @param random_state - A seed value for the random engine
     */
    function DecisionTreeClassifier(_a) {
        var _b = _a === void 0 ? {
            featureLabels: null,
            verbose: false,
            random_state: null,
        } : _a, _c = _b.featureLabels, featureLabels = _c === void 0 ? null : _c, _d = _b.verbose, verbose = _d === void 0 ? false : _d, _e = _b.random_state, random_state = _e === void 0 ? null : _e;
        this.featureLabels = null;
        this.tree = null;
        this.verbose = true;
        this.randomState = null;
        this.randomEngine = null;
        this.featureLabels = featureLabels;
        this.verbose = verbose;
        this.randomState = random_state;
        if (!Number.isInteger(random_state)) {
            this.randomEngine = Random.engines.mt19937().autoSeed();
        }
        else {
            this.randomEngine = Random.engines.mt19937().seed(random_state);
        }
    }
    /**
     * Fit date, which builds a tree
     * @param {any} X - 2D Matrix of training
     * @param {any} y - 1D Vector of target
     * @returns {Leaf | DecisionNode}
     */
    DecisionTreeClassifier.prototype.fit = function (X, y) {
        if (X === void 0) { X = null; }
        if (y === void 0) { y = null; }
        validation_1.validateFitInputs(X, y);
        this.tree = this.buildTree({ X: X, y: y });
    };
    /**
     * Predict multiple rows
     *
     * @param X - 2D Matrix of testing data
     */
    DecisionTreeClassifier.prototype.predict = function (X) {
        if (X === void 0) { X = []; }
        validation_1.validateMatrix2D(X);
        var result = [];
        for (var i = 0; i < X.length; i++) {
            var row = X[i];
            result.push(this._predict({ row: row, node: this.tree }));
        }
        return result;
    };
    /**
     * Returns the model checkpoint
     * @returns {{featureLabels: string[]; tree: any; verbose: boolean}}
     */
    DecisionTreeClassifier.prototype.toJSON = function () {
        return {
            featureLabels: this.featureLabels,
            tree: this.tree,
            verbose: this.verbose,
            random_state: this.randomState,
        };
    };
    /**
     * Restores the model from a checkpoint
     * @param {string[]} featureLabels - Literal names for each feature to be used while printing the tree out as a string
     * @param {any} tree - The model's state
     * @param {boolean} verbose - Logs the progress of the tree construction as console.info
     * @param {number} random_state - A seed value for the random engine
     */
    DecisionTreeClassifier.prototype.fromJSON = function (_a) {
        var _b = _a.featureLabels, featureLabels = _b === void 0 ? null : _b, _c = _a.tree, tree = _c === void 0 ? null : _c, _d = _a.verbose, verbose = _d === void 0 ? false : _d, _e = _a.random_state, random_state = _e === void 0 ? null : _e;
        this.featureLabels = featureLabels;
        this.tree = tree;
        this.verbose = verbose;
        this.randomState = random_state;
    };
    /**
     * Recursively print the tree into console
     * @param {string} spacing - Spacing used when printing the tree into the terminal
     */
    DecisionTreeClassifier.prototype.printTree = function (spacing) {
        if (spacing === void 0) { spacing = ''; }
        if (!this.tree) {
            throw new Error('You cannot print an empty tree');
        }
        this._printTree({ node: this.tree, spacing: spacing });
    };
    /**
     * Partition X and y into true and false branches
     * @param X
     * @param y
     * @param {Question} question
     * @returns {{trueX: Array<any>; trueY: Array<any>; falseX: Array<any>; falseY: Array<any>}}
     */
    DecisionTreeClassifier.prototype.partition = function (X, y, question) {
        var trueX = [];
        var trueY = [];
        var falseX = [];
        var falseY = [];
        for (var i = 0; i < X.length; i++) {
            var row = X[i];
            if (question.match(row)) {
                trueX.push(X[i]);
                trueY.push(y[i]);
            }
            else {
                falseX.push(X[i]);
                falseY.push(y[i]);
            }
        }
        return { trueX: trueX, trueY: trueY, falseX: falseX, falseY: falseY };
    };
    /**
     * Calculate the gini impurity of rows
     * Checkout: https://en.wikipedia.org/wiki/Decision_tree_learning#Gini_impurity
     * @param targets
     * @returns {number}
     */
    DecisionTreeClassifier.prototype.gini = function (targets) {
        var counts = classCounts(targets);
        var impurity = 1;
        var keys = Object.keys(counts);
        for (var i = 0; i < keys.length; i++) {
            var key = keys[i];
            var count = counts[key].count;
            if (count === null || count === undefined) {
                throw Error('Invalid class count detected!');
            }
            var probOfClass = count / targets.length;
            impurity -= Math.pow(probOfClass, 2);
        }
        return impurity;
    };
    /**
     * Information Gain.
     *
     * The uncertainty of the starting node, minus the weighted impurity of
     * two child nodes.
     * @param left
     * @param right
     * @param uncertainty
     * @returns {number}
     */
    DecisionTreeClassifier.prototype.infoGain = function (left, right, uncertainty) {
        var p = left.length / (left.length + right.length);
        return uncertainty - p * this.gini(left) - (1 - p) * this.gini(right);
    };
    /**
     * Find the best split for the current X and y.
     * @param X
     * @param y
     * @returns {{bestGain: number; bestQuestion: any}}
     */
    DecisionTreeClassifier.prototype.findBestSplit = function (X, y) {
        var uncertainty = this.gini(y);
        var nFeatures = X[0].length;
        var bestGain = 0;
        var bestQuestion = null;
        var featureIndex = [];
        if (Number.isInteger(this.randomState)) {
            // method 1: Randomly selecting features
            while (featureIndex.length <= nFeatures) {
                var index = Random.integer(0, nFeatures)(this.randomEngine);
                featureIndex.push(index);
            }
        }
        else {
            featureIndex = lodash_1.range(0, X[0].length);
        }
        var _loop_1 = function (i) {
            var col = featureIndex[i];
            var uniqFeatureValues = lodash_1.uniqBy(lodash_1.map(X, function (row) { return row[col]; }), function (x) { return x; });
            for (var j = 0; j < uniqFeatureValues.length; j++) {
                var feature = uniqFeatureValues[j];
                // featureLabels is for the model interoperability
                var question = new Question(this_1.featureLabels, col, feature);
                // Try splitting the dataset
                var _a = this_1.partition(X, y, question), trueY = _a.trueY, falseY = _a.falseY;
                // Skip this dataset if it does not divide
                if (trueY.length === 0 || falseY.length === 0) {
                    continue;
                }
                // Calculate information gained from this split
                var gain = this_1.infoGain(trueY, falseY, uncertainty);
                if (this_1.verbose) {
                    console.info("fn: " + col + " fval: " + feature + " gini: " + gain);
                }
                if (gain >= bestGain) {
                    bestGain = gain;
                    bestQuestion = question;
                }
            }
        };
        var this_1 = this;
        for (var i = 0; i < featureIndex.length; i++) {
            _loop_1(i);
        }
        return { bestGain: bestGain, bestQuestion: bestQuestion };
    };
    /**
     * Interactively build tree until it reaches the terminal nodes
     * @param {any} X
     * @param {any} y
     * @returns {any}
     */
    DecisionTreeClassifier.prototype.buildTree = function (_a) {
        var X = _a.X, y = _a.y;
        var _b = this.findBestSplit(X, y), bestGain = _b.bestGain, bestQuestion = _b.bestQuestion;
        if (bestGain === 0) {
            return new Leaf(y);
        }
        // Partition the current passed in X ,y
        var _c = this.partition(X, y, bestQuestion), trueX = _c.trueX, trueY = _c.trueY, falseX = _c.falseX, falseY = _c.falseY;
        // Recursively build the true branch
        var trueBranch = this.buildTree({ X: trueX, y: trueY });
        // Recursively build the false branch
        var falseBranch = this.buildTree({ X: falseX, y: falseY });
        return new DecisionNode(bestQuestion, trueBranch, falseBranch);
    };
    /**
     * Internal predict method separated out for recursion purpose
     * @param {any} row
     * @param {any} node
     * @returns {any}
     * @private
     */
    DecisionTreeClassifier.prototype._predict = function (_a) {
        var row = _a.row, node = _a.node;
        if (node instanceof Leaf) {
            // Just return the highest voted
            return node.prediction;
        }
        if (node.question.match(row)) {
            return this._predict({ row: row, node: node.trueBranch });
        }
        else {
            return this._predict({ row: row, node: node.falseBranch });
        }
    };
    /**
     * Private method for printing tree; required for recursion
     * @param {any} node
     * @param {any} spacing
     */
    DecisionTreeClassifier.prototype._printTree = function (_a) {
        var node = _a.node, _b = _a.spacing, spacing = _b === void 0 ? '' : _b;
        if (node instanceof Leaf) {
            console.info(spacing + '' + node.prediction);
            return;
        }
        // Print the question of the node
        console.info(spacing + node.question.toString());
        // Call this function recursively for true branch
        console.info(spacing, '--> True');
        this._printTree({ node: node.trueBranch, spacing: spacing + ' ' });
        // Call this function recursively for false branch
        console.info(spacing, '--> False');
        this._printTree({ node: node.falseBranch, spacing: spacing + ' ' });
    };
    return DecisionTreeClassifier;
}());
exports.DecisionTreeClassifier = DecisionTreeClassifier;
